perceval/backends/github.py

```diff
@@ -0,0 +1,250 @@
+        :param token: GitHub auth token to access the API
+        :param cache: Use issues already retrieved in cache
+        """
+        super().__init__(cache=cache)
+        self.owner = owner
+        self.repository = repository
+        self.client = GitHubClient(owner, repository, token)
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the issues from the repository.
+
+        The method retrieves, from a GitHub repository, the issues
+        updated since the given date.
+
+        :param from_date: obtain issues updated since this date
+
+        :returns: a generator of issues
+        """
+
+        self._purge_cache_queue()
+
+        issues = self.client.get_issues(from_date)
+
+        self._push_cache_queue(issues)
+        self._flush_cache_queue()
+
+        while issues:
+            issue = issues.pop(0)
+            yield issue
+
+            if not issues:
+                issues = self.client.get_issues()
+                if issues and len(issues)>0:
+                    self._push_cache_queue(issues)
+                    self._flush_cache_queue()
+
+    def fetch_from_cache(self):
+        """Fetch the issues from the cache.
+
+        It returns the issues stored in the cache object provided during
+        the initialization of the object. If this method is called but
+        no cache object was provided, the method will raise a `CacheError`
+        exception.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        for issues in cache_items:
+            for issue in issues:
+                yield issue
+
+
+class GitHubClient:
+
+    def __init__(self, owner, repository, token):
+        self.owner = owner
+        self.repository = repository
+        self.auth_token = token
+        self.last_page = self.page = 1  # pagination in items downloading
+
+    def _get_url(self):
+        github_api = "https://api.github.com"
+        github_api_repos = github_api + "/repos"
+        url_repo = github_api_repos + "/" + self.owner + "/" + self.repository
+        return url_repo
+
+    def _get_issues_url(self, startdate=None):
+        # 100 in other items. 20 for pull requests. 30 issues
+        github_per_page = 30
+
+        url_issues = self._get_url() + "/issues"
+
+        url_params = "?per_page=" + str(github_per_page)
+        url_params += "&state=all"  # open and close pull requests
+        url_params += "&sort=updated"  # sort by last updated
+        url_params += "&direction=asc"  # first older pull request
+        if startdate:
+            startdate = startdate.isoformat()
+            url_params += "&since=" + startdate
+
+        url = url_issues + url_params
+
+        return url
+
+    def get_issues(self, start=None):
```

Comment:
In my opinion, this should work like a generator. Thus, multiple calls can be launched at the same time using the same client.